### PR TITLE
Non-blocking Next method for C++ async completion queue

### PR DESCRIPTION
**Pull request by vjpai**
Code changes required to address issue #1004 .

### CODE REVIEW

**Comment by vjpai** [comment]
Do not merge this yet as there are no tests for the new non-blocking method.

File: include/grpc++/completion_queue.h [hunk 1]
```diff
@@ -75,10 +76,18 @@ class CompletionQueue {
...
-  // Blocking read from queue.
-  // Returns true if an event was received, false if the queue is ready
-  // for destruction.
-  bool Next(void **tag, bool *ok);
+  // Tri-state return for AsyncNext: SHUTDOWN, GOT_EVENT, TIMEOUT
+  enum NextStatus {SHUTDOWN, GOT_EVENT, TIMEOUT};
+
+  // Nonblocking (until deadline) read from queue.
+  // Cannot rely on result of tag or ok if return is TIMEOUT
+  NextStatus AsyncNext(void **tag, bool *ok, gpr_timespec deadline);
```
Comment (ctiller) [inline include/grpc++/completion_queue.h]: We should probably use std chrono here instead of the gpr type.

Comment (nicolasnoble) [inline include/grpc++/completion_queue.h]: I'm not entirely sure std::chrono was there in VS2010... At least I don't see the header file in its STL.

Comment (vjpai) [inline include/grpc++/completion_queue.h]: https://msdn.microsoft.com/en-us/library/hh874757.aspx confirms that this is a VS12+ thing. If our C++11 standard is based on VS10, then it's a no-go. I realize that that decision would affect some tests, as some tests are already using std::chrono.

Comment (ctiller) [inline include/grpc++/completion_queue.h]: We've already got code using std::chrono elsewhere - which is why I made
the suggestion. I realize it's out for 2010, but the solution we choose
there will be pretty much a copy/paste when we choose it. My main point was
that it should _not_ be gpr_timespec - we've got good hygiene presently at
not exposing grpc/ headers in grpc++/, and I think that's worthwhile
sticking to.

Suggest using std::chrono now, and when we do 2010, defining some minimal
compatible C++ type for use there.

**Comment by vjpai** [comment]
Ok, I think that you're right. I will change the gpr_timespec uses here to std::chrono . I also have some tests that will get committed and pushed today.


- Vijay

**Comment by vjpai** [comment]
I incorrectly classified the last commit as "currently buggy". It looks like it was ok after all (first try!). I believe that this is now ready for review.

**Comment by vjpai** [comment]
I amended the commit to correct the problem in the commit message.

**Comment by vjpai** [comment]
Thanks for the merge!

**Pull request merged**